xdsl/dialects/linalg.py

```diff
@@ -103,6 +105,45 @@ def __init__(
             regions=[body],
         )
 
+    def get_num_loops(self) -> int:
+        return self.indexing_maps.data[0].data.num_dims
+
+    def get_loops_to_shapes_map(self) -> AffineMap:
+        result_exprs: list[AffineExpr] = []
+        for map_attr in self.indexing_maps:
+            map = map_attr.data
+            result_exprs.extend(map.results)
+
+        dims = self.get_num_loops()
+        # FIXME: Check for symbols also. Currently, we do not allow dynamic
+        # shapes so this is not possible.
+        syms = 0
+
+        return AffineMap(dims, syms, result_exprs)
+
+    def get_shapes_to_loops_map(self) -> AffineMap:
+        loops_to_shapes = self.get_loops_to_shapes_map()
+        inverse = loops_to_shapes.inverse_permutation()
+        if not inverse:
+            raise NotImplementedError("Non-invertible map")
```

Comment:
Is this actually not implemented, or is it just impossible to get it if the permutation is not invertible?

Reply:
If the map is not invertible, Linalg is supposed to use dynamic shapes for it. Since I do not have any support for dynamic shapes in Linalg, this is not implemented yet.

I don't think there will be support for this anytime soon, since I plan to focus on static shapes only.

Reply:
Maybe give a more descriptive error then?
```suggestion
            raise NotImplementedError("Non-invertible maps need dynamic shapes, whicha re not yet implemented.")
```